Documentation/proposal-runconfig_inside_osv.md

```diff
@@ -1,55 +1,286 @@
+
+It must be as easy as possible to add new runtime since we may expect PRs for new runtimes!
+
+QUICK THOUGHT: Is it really neccessary that runtime is hard-coded in Capstan? Would it be better to
+support runtimes written in yaml and included in package itself? Not sure this is possible, though.
+
+##### Package developer
+Statically compiles new languages from source into shared objects and bundles them into Capstan
+package.
+
+NOTE: There is a mess with the hierarchy of packages. Would it make sense to introduce "package type"?
+We could have:
+ * *shared-object*: Packages that only provide shared objects
+ * *runtime*: Packages that provide runtime (may require multiple *shared-object*). These are
+ packages that doesn't work per-se; user needs to provide application code to compose with.
+ * *application*: Packages that user should just download, compose without modifications and run.
+    Only for this packages it's possible to provide the already-compiled image. These packages must
+    make use of Agent functionality.
+
+Role 'Package developer' is the one that prepares *shared-object* packages, role 'Runtime developer'
+prepares *runtime* packages and role 'Runtime user' prepares *application* packages. Role
+'Application user' depends on Runtime user to publish their application.
+
+
+
 ### Composing image
-File meta/run.yaml is present at package root directory (same as until now)
+When user runs `capstan package compose`, the content of meta/run.yaml file is processed and new
+files are generated and uploaded to the OSv image (assume we upload them to folder `/osv-runtime`):
 
-When user runs `capstan package compose` the content of meta/run.yaml file is processed and new files are generated inside the OSv image folder `/osv-runtime`:
+```
+user-friendly attributes of meta/run.yaml ---> complex run command for each named config
+```
 
-Capstan should create a separate file for each named configuration e.g.
+A separate file containing only run command string is created for each named configuration e.g.
 ```
 /osv-runtime
    |--burek
-   |--pizza   <--------\
-   |                    |
-   |--default-runtime  /
+   |--pizza
 ```
-Each runtime file should contain only one line representing the run command that is to be run inside OSv:
+Example:
 ```
 # content of file /osv-runtime/burek:
 node burek.js
 ```
+It is neccessary to use Capstan to generate run command since runtime user does not have enough
+information to generate it on her own (e.g. what binariy to call, what flags to use). The idea
+is that runtime user provides application specific data (e.g. what is her main file) and the rest is
+automatized.
+
+#### Placeholders
+Generation of capstan run command will be updated with placeholder capability.
+If a *value* in meta/run.yaml will start with dollar sign *$* it will be interpreted as placeholder
+name. E.g.
+```yaml
+# meta/run.yaml
+
+runtime: node
+
+main: app.js
+my_ip: $IP
+output_dir: $THE_OUTPUT_DIR
```

Comment:
`my_ip` and `output` are now part of the runtime configuration? How are they linked to the runtime itself? I mean, who knows that they need to be added to the corresponding `node` command as `-ip` and `-out`?